src/cart/crt_hg.c

```diff
@@ -858,6 +1087,48 @@ crt_rpc_handler_common(hg_handle_t hg_hdl)
 	D_ASSERT(proc != NULL);
 	opc = rpc_tmp.crp_req_hdr.cch_opc;
 
+	if (crt_gdata.cg_na_plugin == CRT_NA_OFI_PSM2) {
+		/** Search for the recieved hg_addr in the ha hash table. If
+ 		 *  not found, then insert.
+	 	 */
+        	halink = d_hash_rec_find(&ha_hash_table,
+                	        (void *)&rpc_tmp.crp_req_hdr.cch_clid, 
+				sizeof(rpc_tmp.crp_req_hdr.cch_clid));
+        	if (halink != NULL) {
+			D_DEBUG(DB_TRACE, "client id %" PRIu64 "already in hash table\n",
+				rpc_tmp.crp_req_hdr.cch_clid);
+
+        		d_hash_rec_decref(&ha_hash_table, halink);
+        	}
+		else {
+			ha_entry.hg_class = hg_info->hg_class;	
+			HG_Addr_dup(hg_info->hg_class, hg_info->addr, &ha_entry.hg_addr);
+       			ha = crt_ha_mapping_init(rpc_tmp.crp_req_hdr.cch_clid, ha_entry);
+        		if (!ha) {
+                		D_ERROR("Failed to allocate entry\n");
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs